### PR TITLE
Add DFP, comments page banners, and in feed programatic fallbacks

### PR DESCRIPTION
**Pull request by griffinmichl**
👓 @dwick

### CODE REVIEW

File: src/app/components/BannerAd/index.jsx [hunk 1]
```diff
@@ -0,0 +1,129 @@
+import React from 'react';
+import { connect } from 'react-redux';
+import { createStructuredSelector } from 'reselect';
+import classNames from 'lib/classNames';
+import getSubreddit from 'lib/getSubredditFromState';
+import createBannerProperties from 'lib/createBannerProperties';
+import { defineSlot, destroySlot, getSlotId } from 'lib/dfp';
+import './style.less';
+
+function deferUntilMount(Comp) {
```
Comment (dwick) [inline src/app/components/BannerAd/index.jsx]: you should put this in a separate file. seems like it could be useful elsewhere

File: src/app/components/BannerAd/index.jsx [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        }) }
+      >
+        <div ref={ ref => { this.frame = ref; } } id={ id }></div>
+      </div>
+    );
+  }
+}
+
+const subredditSelector = (state) => {
+  // const subredditName = isFakeSubreddit(listingName) ? '' : listingName.toLowerCase();
```
Comment (dwick) [inline src/app/components/BannerAd/index.jsx]: remove

File: src/app/components/BannerAd/index.jsx [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+      }
+
+      return <Comp { ...this.props } />;
+    }
+  };
+}
+
+
+
+class BannerAd extends React.Component {
```
Comment (dwick) [inline src/app/components/BannerAd/index.jsx]: can you define prop types for this?

File: src/lib/adLocationForPostRecords.js [hunk 4]
```diff
@@ -3,3 +3,18 @@ import { AD_LOCATION } from 'app/constants';
 export default function(postRecords=[]) {
   return Math.min(AD_LOCATION, postRecords.length);
 }
+
+const canShowAdNextTo = post => !post.over18 && post.brandSafe;
+
+export const dfpAdLocationFromPosts = (posts = []) => {
```
Comment (dwick) [inline src/lib/adLocationForPostRecords.js]: 👍

File: src/lib/dfp.js [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+  if (isFakeSubreddit(subredditName)) {
+    parts.push(FRONTPAGE_NAME);
+  } else {
+    parts.push(`r.${subredditName}`);
+  }
+
+  parts.push(listingName);
+
+  // We don't care about distinguishing the index route
```
Comment (dwick) [inline src/lib/dfp.js]: i think this is specific to reddit-desktop? can you just remove it

File: src/app/components/BannerAd/style.less [hunk 6]
```diff
@@ -0,0 +1,15 @@
+@import (reference) '~app/less/themes/themeify';
+
+.BannerAd {
+  &__320x50 {
```
Comment (dwick) [inline src/app/components/BannerAd/style.less]: I think you may want to set a height for this as well to prevent it from reflowing when the ad comes in.

File: src/platform/Server.js [hunk 7]
```diff
@@ -108,7 +108,7 @@ export default config => {
   server.use(router.allowedMethods());
 
   return () => {
-    server.listen(port, () => {
+    server.listen(port, '0.0.0.0', () => {
```
Comment (dwick) [inline src/platform/Server.js]: this should be a separate commit

File: src/app/components/BannerAd/style.less (hunk 6 above)
Comment (griffinmichl) [inline src/app/components/BannerAd/style.less]: I've configured DFP to hold the space available for the ad until the request is fulfilled. It then collapses if the slot could not be filled.

File: src/platform/Server.js (hunk 7 above)
Comment (griffinmichl) [inline src/platform/Server.js]: Woops. That shouldn't even be in there

**Comment by griffinmichl** [comment]
💅

**Comment by griffinmichl** [comment]
Wait no 💇

**Comment by dwick** [comment]
🏃
